### PR TITLE
Alternate Solution to CSES 1621 - Distinct Numbers

### PR DESCRIPTION
**Pull request by dutinmeow**

### CODE REVIEW

**Comment by dutinmeow** [comment]
I just saw the set solution in the Introduction to Sets and Maps module, but the alternate solution should still be on the solution page

**Comment by maggie-j-liu** [comment]
Hmm i don't know if duplicating solutions is a good idea, but can you copy paste the java and python code too?

**Comment by dutinmeow** [comment]
> Hmm i don't know if duplicating solutions is a good idea, but can you copy paste the java and python code too?

Done, although I'm not sure who wrote the Java or Python codes so I can't credit them

**Comment by bqi343** [comment]
> I just saw the set solution in the Introduction to Sets and Maps module, but the alternate solution should still be on the solution page

why not just link to the module?

**Comment by dutinmeow** [comment]
I've added the link to the module in the solution, but I think it's still nice to have all the methods of solving a problem within the solution page

**Comment by bqi343** [comment]
personally, not duplicating material takes priority over that. let's see what others think

File: solutions/cses-1621.mdx [hunk 1]
```diff
@@ -96,3 +96,84 @@ print(ans)
 </PySection>
 
 </LanguageSection>
+
+## Alternative Solution
+	
+By using an std::set in C++ or its equivalents in other languages, we can take advantage of the fact that a set only stores a copy of an element once.
+Sets and Maps are introduced [here](https://usaco.guide/bronze/intro-sets?lang=py). Furthermore, using (an unordered set)[https://usaco.guide/gold/unordered?lang=py] will enable an $\mathcal{O}(n)$ complexity, although $\texttt{chash}$ is necessary to avoid a TLE verdict.
```
Comment (bqi343) [inline solutions/cses-1621.mdx:103]: why `lang=py`?

File: solutions/cses-1621.mdx [hunk 2]
```diff
@@ -96,3 +96,84 @@ print(ans)
 </PySection>
 
 </LanguageSection>
+
+## Alternative Solution
+	
+By using an std::set in C++ or its equivalents in other languages, we can take advantage of the fact that a set only stores a copy of an element once.
```
Comment (bqi343) [inline solutions/cses-1621.mdx:102]:
```suggestion
By using an `std::set` in C++ or its equivalents in other languages, we can take advantage of the fact that a set only stores a copy of an element once.
```

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/cses-1621.mdx [hunk 3]
```diff
@@ -96,3 +96,84 @@ print(ans)
...
+<JavaSection>
+
+```java
+// Source: Daniel
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.HashSet;
+import java.util.StringTokenizer;
```
Comment (sathu-beep) [inline solutions/cses-1621.mdx:139]:
```suggestion
import java.io*;
import java.util*;
```
There is no reason not to do this it makes the code much nicer to look at

File: solutions/cses-1621.mdx [hunk 4]
```diff
@@ -96,3 +96,84 @@ print(ans)
...
+		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+		int n = Integer.parseInt(br.readLine());
+		StringTokenizer st = new StringTokenizer(br.readLine());
+		HashSet<Integer> set = new HashSet<Integer>();
+		for (int i = 0; i < n; i++) {
+			set.add(Integer.parseInt(st.nextToken()));
+		}
+		System.out.println(set.size());
+	}
+
```
Comment (sathu-beep) [inline solutions/cses-1621.mdx:153]:
```suggestion
```
whitespace 😒

File: solutions/cses-1621.mdx [hunk 5]
```diff
@@ -96,3 +96,84 @@ print(ans)
...
+print(len(distinct_nums))
+```
+
+We can do this more efficiently by skipping the creation of the list, and use a
+set comprehension directly:
+
+```py
+n = int(input()) # unused
+distinct_nums = {int(x) for x in input().split()}
+print(len(distinct_nums))
```
Comment (sathu-beep) [inline solutions/cses-1621.mdx:174]:
```suggestion
```py
n = int(input()) # unused
distinct_nums = {int(x) for x in input().split()}
print(len(distinct_nums))
```

just implement the more efficient solution 
its not that much different than the first that it deserves to be an entire new block of code

Comment (sathu-beep) [reply solutions/cses-1621.mdx:174]: oops i think that the ``` at the end of the python code were removed when I formatted this comment make sure to add them back

**Review by sathu-beep — CHANGES_REQUESTED** [review]
address formatting issues

File: solutions/cses-1621.mdx (hunk 5 above)
Comment (sathu-beep) [reply solutions/cses-1621.mdx:174]: @SansPapyrus683 😏 i vow to be more nitpicky than you in the future

**Comment by bqi343** [comment]
@maggie-j-liu what do you think? do you want to replace this w/ a link to the module? 

https://usaco.guide/bronze/intro-sets#solution---distinct-numbers

File: solutions/cses-1621.mdx [hunk 6]
```diff
@@ -96,3 +96,70 @@ print(ans)
...
+<JavaSection>
+
+```java
+// Source: Daniel
+
+import java.io*;
+import java.util*;
+
+public class DistinctNumbers {
+
```
Comment (sathu-beep) [inline solutions/cses-1621.mdx:139]:
```suggestion
```
whitespace

File: solutions/cses-1621.mdx [hunk 7]
```diff
@@ -96,3 +96,70 @@ print(ans)
...
 </LanguageSection>
+
+## Alternative Solution
+	
+By using an `std::set` in C++ or its equivalents in other languages, we can take advantage of the fact that a set only stores a copy of an element once.
+Sets and Maps are introduced [here](https://usaco.guide/bronze/intro-sets). Furthermore, using (an unordered set)[https://usaco.guide/gold/unordered] will enable an $\mathcal{O}(n)$ complexity, although $\texttt{chash}$ is necessary to avoid a TLE verdict. 
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log{N})$
+
```
Comment (sathu-beep) [inline solutions/cses-1621.mdx:107]:
```suggestion
```

File: solutions/cses-1621.mdx [hunk 8]
```diff
@@ -96,3 +96,70 @@ print(ans)
...
+int main() {
+	cin >> N;
+	for (int i = 0; i < N; i++) {
+		int a; cin >> a;
+		S.insert(a);
+	}
+	cout << S.size() << '\n';
+}
+```
+</CPPSection>
```
Comment (sathu-beep) [inline solutions/cses-1621.mdx:129]:
```suggestion
</CPPSection>

```

**Comment by maggie-j-liu** [comment]
> @maggie-j-liu what do you think? do you want to replace this w/ a link to the module?
> 
> https://usaco.guide/bronze/intro-sets#solution---distinct-numbers

Yeah, I think just linking to the module is fine, since it's unnecessary to copy code from one place to another

**Review by sathu-beep — COMMENTED** [review]
format

**Comment by bqi343** [comment]
see #1833

**Pull request closed without merging**